llama-rs/src/lib.rs

```diff
@@ -1738,3 +1745,41 @@ impl Vocabulary {
         Ok(res)
     }
 }
+
+/// Used to buffer incoming tokens until they produce a valid string of UTF-8 text.
+///
+/// Tokens are *not* valid UTF-8 by themselves. However, the LLM will produce valid UTF-8
+/// from multiple tokens. This helps alleviate that issue.
+#[derive(Clone, PartialEq, Default)]
+pub struct TokenUtf8Buffer(Vec<u8>);
+impl TokenUtf8Buffer {
+    /// Create a new buffer.
+    pub const fn new() -> Self {
+        Self(vec![])
+    }
+
+    /// Add a token to the buffer. If the buffer contains a valid string of UTF-8 text,
+    /// it is returned and the buffer is cleared for next use.
```

Comment:
One thing I'm not sure about this design is what would happen if invalid UTF-8 ends up being in the stream somehow. After all, the LLM is not 100% reliable in that regard, so if it uses the byte fallback tokens and produces an invalid sequence of bytes, `str::from_utf8` will fail from there onwards, and there doesn't seem to be a way to get out of that situation. From the user end, no more tokens would be produced.

I think one way to solve this would be to attempt an error recovery strategy that skips any error bytes until it finds a substring that is valid utf-8. That is, try parsing the slice after dropping the first i tokens, for an i ranging from 0 to `&self.0.len() - 2`. If any `i` parses correctly, any previous bytes should be printed as replacement characters (�) instead. This has some "accidentally quadratic" potential, but I think in practice the sequences will be too short for this to matter.

I'm not an expert in how UTF-8 works, but I think this behaviour would be correct, because in UTF-8 no character is a prefix of any other character? But I'm not 100% sure on this.